File: step_definitions/Verify user can adjust product quantities in the cart.java
package com.cucumber.steps;

import com.page_objects.CartPage;
import io.cucumber.java.en.Given;
import io.cucumber.java.en.Then;
import io.cucumber.java.en.When;

public class CartManagementSteps extends CartPage {

    @Given("the user is logged into the application")
    public void theUserIsLoggedIntoTheApplication() {
        loginToApplication();
    }

    @Given("at least one product is added to the cart")
    public void atLeastOneProductIsAddedToTheCart() {
        addProductToCart();
    }

    @When("the user navigates to the shopping cart page")
    public void theUserNavigatesToTheShoppingCartPage() {
        openCart();
    }

    @Then("the shopping cart page is displayed with current products")
    public void theShoppingCartPageIsDisplayedWithCurrentProducts() {
        verifyCartPageDisplayed();
    }

    @Given("the product with Product ID: {int} is visible in the cart")
    public void theProductWithProductIDIsVisibleInTheCart(int productId) {
        verifyProductInCart(productId);
    }

    @When("the user enters a new quantity in the quantity field")
    public void theUserEntersANewQuantityInTheQuantityField() {
        enterNewQuantity();
    }

    @Then("the quantity field accepts the new value")
    public void theQuantityFieldAcceptsTheNewValue() {
        verifyQuantityFieldAcceptsValue();
    }

    @Given("the user has entered a new quantity")
    public void theUserHasEnteredANewQuantity() {
        enterNewQuantity();
    }

    @When("the user clicks the 'Update' button")
    public void theUserClicksTheUpdateButton() {
        clickUpdateButton();
    }

    @Then("the product quantity is updated in the cart")
    public void theProductQuantityIsUpdatedInTheCart() {
        verifyProductQuantityUpdated();
    }

    @Then("a confirmation message is displayed")
    public void aConfirmationMessageIsDisplayed() {
        verifyConfirmationMessageDisplayed();
    }

    @Given("the product quantity is updated")
    public void theProductQuantityIsUpdated() {
        verifyProductQuantityUpdated();
    }

    @When("the user checks the total price")
    public void theUserChecksTheTotalPrice() {
        checkTotalPrice();
    }

    @Then("the total price reflects the new quantity")
    public void theTotalPriceReflectsTheNewQuantity() {
        verifyTotalPriceReflectsQuantity();
    }

    @Given("the user attempts to set the quantity to zero")
    public void theUserAttemptsToSetTheQuantityToZero() {
        attemptToSetQuantityToZero();
    }

    @Then("the system prevents setting quantity to zero")
    public void theSystemPreventsSettingQuantityToZero() {
        verifySystemPreventsZeroQuantity();
    }

    @Then("displays an error message")
    public void displaysAnErrorMessage() {
        verifyErrorMessageDisplayed();
    }

    @Given("the user enters a non-numeric value in the quantity field")
    public void theUserEntersANonNumericValueInTheQuantityField() {
        enterNonNumericValue();
    }

    @Then("the system displays an error message indicating invalid input")
    public void theSystemDisplaysAnErrorMessageIndicatingInvalidInput() {
        verifyInvalidInputErrorMessage();
    }

    @Given("the user increases the quantity to a number greater than available stock")
    public void theUserIncreasesTheQuantityToANumberGreaterThanAvailableStock() {
        increaseQuantityBeyondStock();
    }

    @Then("the system displays a warning about insufficient stock")
    public void theSystemDisplaysAWarningAboutInsufficientStock() {
        verifyInsufficientStockWarning();
    }

    @Given("the user decreases the quantity to a lower number")
    public void theUserDecreasesTheQuantityToALowerNumber() {
        decreaseQuantity();
    }

    @Then("the quantity is updated successfully")
    public void theQuantityIsUpdatedSuccessfully() {
        verifyQuantityUpdatedSuccessfully();
    }

    @Then("the total price is recalculated")
    public void theTotalPriceIsRecalculated() {
        verifyTotalPriceRecalculated();
    }

    @Given("the user changes the quantity")
    public void theUserChangesTheQuantity() {
        changeQuantity();
    }

    @Then("the cart reflects the updated quantity immediately")
    public void theCartReflectsTheUpdatedQuantityImmediately() {
        verifyCartReflectsUpdatedQuantity();
    }

    @Given("the user logs out and logs back in")
    public void theUserLogsOutAndLogsBackIn() {
        logoutAndLogin();
    }

    @Then("the cart retains updated quantities after re-login")
    public void theCartRetainsUpdatedQuantitiesAfterReLogin() {
        verifyCartPersistenceAfterReLogin();
    }

    @Given("the user adds a new product to the cart")
    public void theUserAddsANewProductToTheCart() {
        addNewProductToCart();
    }

    @When("the user adjusts its quantity")
    public void theUserAdjustsItsQuantity() {
        adjustNewProductQuantity();
    }

    @Then("the new product quantity can be adjusted successfully")
    public void theNewProductQuantityCanBeAdjustedSuccessfully() {
        verifyNewProductQuantityAdjusted();
    }

    @Given("the user attempts to adjust quantity for a product with a special promotion")
    public void theUserAttemptsToAdjustQuantityForAProductWithASpecialPromotion() {
        attemptToAdjustPromotionalProductQuantity();
    }

    @Then("quantity adjustment works correctly with promotional products")
    public void quantityAdjustmentWorksCorrectlyWithPromotionalProducts() {
        verifyPromotionalProductQuantityAdjustment();
    }

    @Given("the user changes the quantity of items in the cart")
    public void theUserChangesTheQuantityOfItemsInTheCart() {
        changeItemQuantityInCart();
    }

    @Then("the cart icon displays the correct number of items")
    public void theCartIconDisplaysTheCorrectNumberOfItems() {
        verifyCartIconItemCount();
    }

    @Given("the user checks the system for quantity adjustment for bundled products")
    public void theUserChecksTheSystemForQuantityAdjustmentForBundledProducts() {
        checkBundledProductQuantityAdjustment();
    }

    @Then("the system permits quantity adjustment for bundled products")
    public void theSystemPermitsQuantityAdjustmentForBundledProducts() {
        verifyBundledProductQuantityAdjustment();
    }
}

File: step_definitions/Verify user can adjust product quantities in the cart.js
import { Given, When, Then } from "cypress-cucumber-preprocessor/steps";
import ShoppingCartPage from "../pages/shoppingCartPage";

Given("the user is logged into the application", () => {
 // Implement login functionality
});

Given("at least one product is added to the cart", () => {
 // Implement functionality to add a product to the cart
});

When("the user navigates to the shopping cart page", () => {
 ShoppingCartPage.visit();
});

Then("the shopping cart page is displayed with current products", () => {
 ShoppingCartPage.verifyCartPage();
});

Given("the product with Product ID: {int} is visible in the cart", (productId) => {
 ShoppingCartPage.verifyProductInCart(productId);
});

When("the user enters a new quantity in the quantity field", () => {
 ShoppingCartPage.enterQuantity();
});

Then("the quantity field accepts the new value", () => {
 ShoppingCartPage.verifyQuantityUpdate();
});

Given("the user has entered a new quantity", () => {
 // Implement functionality to enter new quantity
});

When("the user clicks the 'Update' button", () => {
 ShoppingCartPage.clickUpdateButton();
});

Then("the product quantity is updated in the cart", () => {
 ShoppingCartPage.verifyProductQuantityUpdated();
});

Then("a confirmation message is displayed", () => {
 ShoppingCartPage.verifyConfirmationMessage();
});

Given("the product quantity is updated", () => {
 // Implement functionality to verify product quantity update
});

When("the user checks the total price", () => {
 ShoppingCartPage.checkTotalPrice();
});

Then("the total price reflects the new quantity", () => {
 ShoppingCartPage.verifyTotalPriceUpdate();
});

Given("the user attempts to set the quantity to zero", () => {
 // Implement functionality to attempt setting quantity to zero
});

Then("the system prevents setting quantity to zero", () => {
 ShoppingCartPage.verifyPreventZeroQuantity();
});

Then("displays an error message", () => {
 ShoppingCartPage.verifyErrorMessage();
});

Given("the user enters a non-numeric value in the quantity field", () => {
 // Implement functionality to enter non-numeric value
});

Then("the system displays an error message indicating invalid input", () => {
 ShoppingCartPage.verifyInvalidInputMessage();
});

Given("the user increases the quantity to a number greater than available stock", () => {
 // Implement functionality to increase quantity beyond stock
});

Then("the system displays a warning about insufficient stock", () => {
 ShoppingCartPage.verifyInsufficientStockWarning();
});

Given("the user decreases the quantity to a lower number", () => {
 // Implement functionality to decrease quantity
});

Then("the quantity is updated successfully", () => {
 ShoppingCartPage.verifyQuantityUpdateSuccess();
});

Then("the total price is recalculated", () => {
 ShoppingCartPage.verifyTotalPriceRecalculation();
});

Given("the user changes the quantity", () => {
 // Implement functionality to change quantity
});

Then("the cart reflects the updated quantity immediately", () => {
 ShoppingCartPage.verifyRealTimeCartUpdate();
});

Given("the user logs out and logs back in", () => {
 // Implement functionality for logout and login
});

Then("the cart retains updated quantities after re-login", () => {
 ShoppingCartPage.verifyCartPersistenceAfterLogin();
});

Given("the user adds a new product to the cart", () => {
 // Implement functionality to add new product
});

When("the user adjusts its quantity", () => {
 ShoppingCartPage.adjustNewProductQuantity();
});

Then("the new product quantity can be adjusted successfully", () => {
 ShoppingCartPage.verifyNewProductQuantityAdjustment();
});

Given("the user attempts to adjust quantity for a product with a special promotion", () => {
 // Implement functionality for promotional product quantity adjustment
});

Then("quantity adjustment works correctly with promotional products", () => {
 ShoppingCartPage.verifyPromotionalProductQuantityAdjustment();
});

Given("the user changes the quantity of items in the cart", () => {
 // Implement functionality to change item quantity
});

Then("the cart icon displays the correct number of items", () => {
 ShoppingCartPage.verifyCartIconUpdate();
});

Given("the user checks the system for quantity adjustment for bundled products", () => {
 // Implement functionality for bundled product quantity adjustment
});

Then("the system permits quantity adjustment for bundled products", () => {
 ShoppingCartPage.verifyBundledProductQuantityAdjustment();
});